Hw9

EmelyanovPavel opened the pull request:
[homework9.cpp](https://github.com/EmelyanovPavel/HW-9-VAR8-Emelyanov-PM/blob/hw9/homework9.cpp)

Commit:
Задания по рекурсивным функциям из файла пдф.

EmelyanovPavel committed:
Update and rename Organization of file input or output.cpp to homework9

EmelyanovPavel committed:
homework9

EmelyanovPavel committed:
Update homework9.cpp

добавлен текст задания под комментарием "//Task 1"

EmelyanovPavel committed:
Update homework9.cpp

Написан код по заданию 1 и по заданию 2, где выбран 8 пример(вариант), из файла пдф.

tak333 commented on homework9.cpp:
```diff
@@ -0,0 +1,215 @@
+    for (int i = 1; i < n; i++) 
+    {
+        T key = vec[i];
+        int j = i - 1;
```
Используйте тип size_t для векторов.

tak333 replied:
При движении циклом от n до 0 ограничителем будет являться -1.
Будьте аккуратны с условием - иначе может зациклиться.

tak333 commented on homework9.cpp:
```diff
@@ -0,0 +1,215 @@
+
+//Template sorting function with method selection
+template<typename T>
+void sortVector(std::vector<T>& vec, int variant) 
```
У вас вариант 2 по формуле, остальные вам делать было не нужно.

tak333 commented on homework9.cpp:
```diff
@@ -0,0 +1,215 @@
+};
+
+//Function for comparing dates of birth
+bool compareByDate(const Student& a, const Student& b) 
```
Назовите функцию как вопрос, она же возвращает bool: да / нет.

tak333 commented on homework9.cpp:
```diff
@@ -0,0 +1,215 @@
+    }
+
+    // Sorting students by date of birth
+    std::sort(students.begin(), students.end(), compareByDate);
```
Сортировать нужно своей функцией сортировки.

EmelyanovPavel committed:
Update homework9.cpp

Изменён тип данных на "size_t" в условиях операторов "for", где написаны переменные i и j, при работе с векторами; изменено значение варианта для выбора метода сортировки в первой функции войд; переименована функция, возвращающая булево значение.

tak333 commented on homework9.cpp:
```diff
@@ -0,0 +1,219 @@
+    for (size_t i = 1; i < n; i++)
+    {
+        T key = vec[i];
+        int j = i - 1;
```
Для всех индексов используем size_t

tak333 replied:
size_t не добавили для j

tak333 commented on homework9.cpp:
```diff
@@ -0,0 +1,219 @@
+    }
+
+    // Sorting students by date of birth
+    std::sort(students.begin(), students.end(), isMore);
```
Необходимо реализовать свою функцию сортировки студентов.

EmelyanovPavel replied:
Как придумать свою функцию сортировки студентов? Какую свою функцию сортировки студентов нужно реализовать?

tak333 replied:
Реализовать функцию для сортировки вектора студентов тем же методом, что и в первой части задания: пузырьковую, вставками или выбором.

EmelyanovPavel committed:
Update homework9.cpp

Добавлены функции войд с методами сортировки и написана функция войд с выбором метода сортировки по заданию 2.

tak333 commented on homework9.cpp:
```diff
@@ -0,0 +1,289 @@
+        Student key = students[i];
+        size_t j = i - 1;
+
+        while (j >= 0 && !isMore(students[j].birthDate, key.birthDate)) 
```
size_t надо проверять не на отрицательность (оно никогда таким не будет), а проверять не равно ли первому неправильному значению, т.е. -1.

EmelyanovPavel replied:
как это подправить?

EmelyanovPavel committed:
Update homework9.cpp

Написаны все типы size_t при работе с векторами и добавлена проверка на недопустимое значение size_t (-1) во вторую функцию void.

tak333 commented on homework9.cpp:
```diff
@@ -231,19 +231,24 @@ void task2()
-
+    
+    //Checking for an invalid value
+    if (count == static_cast<size_t>(-1)) {
```
Эта проверка сработает только если пользователь введёт строго -1.

tak333 reviewed (changes requested):
Функция сортировки вставками: все индексы должны быть size_t.
Но, тогда вам надо поменять условие остановки, потому что size_t всегда неотрицательный.
Сравнивайте с -1, она преобразуется в другое значение, которое будет эквивалентно встрече -1 на size_t.
